xen/arch/arm/sci/scmi_smc.c

```diff
@@ -0,0 +1,944 @@
+        if ( ret )
+        {
+            printk(XENLOG_ERR "scmi: Could not create channel for agent %d: %d",
+                   agent_id, ret);
+            return ret;
+        }
+
+        printk(XENLOG_DEBUG "scmi: Agent %d SMC %X addr %lx\n", agent_id,
+               smc_id, addr);
+    }
+
+    return 0;
+}
+
+static int scmi_assign_device(struct dt_device_node *dev,
+                              struct dt_phandle_args *ac_spec,
+                              struct domain *d)
+{
+    uint32_t dev_id = ac_spec->args[0];
+
+    return scmi_add_device_by_devid(d, dev_id);
+}
+
+static int scmi_deassign_device(struct dt_device_node *dev,
+                                struct dt_phandle_args *ac_spec,
+                                struct domain *d)
+{
+    return 0;
+}
+
+
+static struct ac_ops scmi_ac_ops =
+{
+    .assign_device = scmi_assign_device,
+    .deassign_device = scmi_deassign_device,
+};
+
+static __init bool scmi_probe(struct dt_device_node *scmi_node)
+{
+    u64 addr, size;
+    int ret, i;
+    struct scmi_channel *channel, *agent_channel;
+    int n_agents;
+    scmi_msg_header_t hdr;
+    struct rx_t {
+        int32_t status;
+        uint32_t attributes;
+    } rx;
+
+    ASSERT(scmi_node != NULL);
+
+    INIT_LIST_HEAD(&scmi_data.channel_list);
+    spin_lock_init(&scmi_data.channel_list_lock);
+
+    if ( !dt_property_read_u32(scmi_node, SCMI_SMC_ID, &scmi_data.func_id) )
+    {
+        printk(XENLOG_ERR "scmi: Unable to read smc-id from DT\n");
+        return false;
+    }
+
+    ret = read_hyp_channel_addr(scmi_node, &addr, &size);
+    if ( IS_ERR_VALUE(ret) )
+        return false;
+
+    if ( !IS_ALIGNED(size, SCMI_SHMEM_MAPPED_SIZE) )
+    {
+        printk(XENLOG_ERR "scmi: Reserved memory is not aligned\n");
+        return false;
+    }
+
+    channel = smc_create_channel(HYP_CHANNEL, scmi_data.func_id, addr);
+    if ( IS_ERR(channel) )
+        goto out;
+
+    ret = map_channel_memory(channel);
+    if ( ret )
+        goto out;
+
+    spin_lock(&scmi_data.channel_list_lock);
+    channel->domain_id = DOMID_XEN;
+    spin_unlock(&scmi_data.channel_list_lock);
+
+    hdr.id = SCMI_BASE_PROTOCOL_ATTIBUTES;
+    hdr.type = 0;
+    hdr.protocol = SCMI_BASE_PROTOCOL;
+
+    printk("%s:%d\n",__func__,  __LINE__);
+    ret = do_smc_xfer(channel, &hdr, NULL, 0, &rx, sizeof(rx));
+    if ( ret )
+        goto error;
+
+    ret = check_scmi_status(rx.status);
+    if ( ret )
+        goto error;
+
+    n_agents = FIELD_GET(MSG_N_AGENTS_MASK, rx.attributes);
+    printk(XENLOG_DEBUG "scmi: Got agent count %d\n", n_agents);
+
+    ret = collect_agents(scmi_node);
+    if ( ret )
+        goto error;
+
+    i = 1;
+
+    list_for_each_entry(agent_channel, &scmi_data.channel_list, list)
+    {
+        uint32_t tx_agent_id = 0xFFFFFFFF;
+        struct {
+            int32_t status;
+            uint32_t agent_id;
+            char name[16];
+        } da_rx;
```

Comment:
I propose to define messages properly at the beginning of the file, like
```
struct scmi_base_discover_agent_p2a {
	int32_t status; <---- GS: this can be avoided.
	uint32_t agent_id;
	char name[SCMI_DEFAULT_STRING_LENGTH];
};
```